Sources/CombineWamp/PeerWorkflow/Client/WampClient.swift

```diff
@@ -10,26 +10,44 @@ import FoundationExtensions
 public struct WampClient {
     let session: WampSession
     let roles: Set<WampRole>
+    let realm: URI
+    let publisherRole: () -> WampPublisherProtocol
+    let subscriberRole: () -> WampSubscriberProtocol
+    let callerRole: () -> WampCallerProtocol
+    let calleeRole: () -> WampCalleeProtocol
 
-    init(session: WampSession, roles: Set<WampRole>) {
+    public init(
+        session: WampSession,
+        roles: Set<WampRole>,
+        realm: URI,
+        publisherRole: @escaping () -> WampPublisherProtocol,
```

Comment:
I'm thinking here that maybe it doesn't make sense to ask for `roles` and the init for all the roles. For example, the caller says that this client only implements `subscriberRole` in the `Set<WampRole>` above, but still, they have to provide constructor for ALL the other 3 roles that this client doesn't implement. Not critical, just strange. Perhaps the WampRole should come with the init, or the argument `roles` wouldn't be passed to the client but instead inferred depending on the 4 role inits you ask (and they would be Optional. As this is not critical, I believe it can be improved in future iterations, just be aware that will break the public contract again.

Reply:
Hm, having this in the Role is a bit much, but we could at least check the params for consistency.
```
public struct WampClient {
    let session: WampSession
    let roles: Set<WampRole>
    let realm: URI
    let publisherRole: (() -> WampPublisherProtocol)?
    let subscriberRole: (() -> WampSubscriberProtocol)?
    let callerRole: (() -> WampCallerProtocol)?
    let calleeRole: (() -> WampCalleeProtocol)?

    public init(
        session: WampSession,
        roles: Set<WampRole>,
        realm: URI,
        publisherRole: (() -> WampPublisherProtocol)? = nil,
        subscriberRole: (() -> WampSubscriberProtocol)? = nil,
        callerRole: (() -> WampCallerProtocol)? = nil,
        calleeRole: (() -> WampCalleeProtocol)? = nil
    ) {
        self.session = session
        self.roles = roles
        self.realm = realm
        self.publisherRole = publisherRole
        self.subscriberRole = subscriberRole
        self.callerRole = callerRole
        self.calleeRole = calleeRole

        roles.forEach { role in
            switch role {
            case .publisher: assert(publisherRole != nil, "You must provide a pubisherRole closure to init, see WampPublisherProtocol.")
            case .subscriber: assert(subscriberRole != nil, "You must provide a subscriberRole closure to init, see WampSubscriberProtocol.")
            case .caller: assert(callerRole != nil, "You must provide a callerRole closure to init, see WampCallerProtocol.")
            case .callee: assert(calleeRole != nil, "You must provide a callerRole closure to init, see WampCalleeProtocol.")
            case .broker:
                break
            case .dealer:
                break
            }
        }
    }
```

This would be the init when we calculate the roles argument from the optional parameters:

```
public struct WampClient {
    let session: WampSession
    let roles: Set<WampRole>
    let realm: URI
    let publisherRole: (() -> WampPublisherProtocol)?
    let subscriberRole: (() -> WampSubscriberProtocol)?
    let callerRole: (() -> WampCallerProtocol)?
    let calleeRole: (() -> WampCalleeProtocol)?

    public init(
        session: WampSession,
        realm: URI,
        publisherRole: (() -> WampPublisherProtocol)? = nil,
        subscriberRole: (() -> WampSubscriberProtocol)? = nil,
        callerRole: (() -> WampCallerProtocol)? = nil,
        calleeRole: (() -> WampCalleeProtocol)? = nil
    ) {
        self.session = session
        self.realm = realm
        self.publisherRole = publisherRole
        self.subscriberRole = subscriberRole
        self.callerRole = callerRole
        self.calleeRole = calleeRole

        var internalRoles: Set<WampRole> = []
        if publisherRole != nil { internalRoles.insert(.publisher) }
        if subscriberRole != nil { internalRoles.insert(.subscriber) }
        if callerRole != nil { internalRoles.insert(.caller) }
        if calleeRole != nil { internalRoles.insert(.callee) }
        self.roles = internalRoles
    }

    public var asPublisher: WampPublisherProtocol? {
        publisherRole?()
    }

    public var asSubscriber: WampSubscriberProtocol? {
        subscriberRole?()
    }

    public var asCaller: WampCallerProtocol? {
        callerRole?()
    }

    public var asCallee: WampCalleeProtocol? {
        calleeRole?()
    }

```

I would prefer the latter version. The `if`s are not great, but map on the optionals would create an unused result which is more confusing.